File: app/src/main/java/com/example/tetrisapp/ui/fragment/CreateLobbyFragment.java
package com.example.tetrisapp.ui.fragment;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import com.example.tetrisapp.R;
import com.example.tetrisapp.data.remote.LobbyService;
import com.example.tetrisapp.databinding.FragmentCreateLobbyBinding;
import com.example.tetrisapp.model.remote.request.CreateLobbyPayload;
import com.example.tetrisapp.model.remote.response.DefaultPayload;
import com.example.tetrisapp.ui.activity.MainActivity;
import com.example.tetrisapp.util.FirebaseTokenUtil;
import com.example.tetrisapp.util.OnTouchListener;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@AndroidEntryPoint
public class CreateLobbyFragment extends Fragment implements Callback<DefaultPayload> {
    private static final String TAG = "CreateLobbyFragment";
    private FragmentCreateLobbyBinding binding;

    @Inject
    FirebaseUser firebaseUser;
    @Inject
    LobbyService lobbyService;

    private Call<DefaultPayload> apiCall;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentCreateLobbyBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initOnClickListeners();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (apiCall != null) apiCall.cancel();
    }

    @SuppressLint("ClickableViewAccessibility")
    private void initOnClickListeners() {
        binding.btnBack.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnBack.setOnClickListener(v -> Navigation.findNavController(binding.getRoot()).popBackStack());

        binding.btnResetSettings.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnResetSettings.setOnClickListener(v -> {
            binding.countdownSlider.setValue(3.0f);
            binding.playerLimitSlider.setValue(2.0f);
            binding.switchEnablePause.setChecked(false);
        });

        binding.btnCreateLobby.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnCreateLobby.setOnClickListener(v -> {
            if (firebaseUser == null) {
                Navigation.findNavController(binding.getRoot()).navigate(R.id.action_createLobbyFragment_to_accountFragment);
                return;
            }

            int countdown = Math.round(binding.countdownSlider.getValue());
            int playerLimit = Math.round(binding.playerLimitSlider.getValue());
            boolean enablePause = binding.switchEnablePause.isChecked();

            FirebaseTokenUtil.getFirebaseToken(idToken -> {
                apiCall = lobbyService.createLobby(new CreateLobbyPayload(idToken, countdown, playerLimit));
                apiCall.enqueue(this);
            });
        });
    }

    @Override
    public void onResponse(@NonNull Call<DefaultPayload> call, Response<DefaultPayload> response) {
        if (response.code() == 401) {
            Navigation.findNavController(binding.getRoot()).navigate(R.id.action_createLobbyFragment_to_accountFragment);
        }

        if (response.isSuccessful() && response.body() != null && response.body().status.equals("success")) {
            CreateLobbyFragmentDirections.ActionCreateLobbyFragmentToLobbyFragment action = CreateLobbyFragmentDirections.actionCreateLobbyFragmentToLobbyFragment();
            action.setInviteCode(response.body().message);
            Navigation.findNavController(binding.getRoot()).navigate(action);
        }
    }

    @Override
    public void onFailure(@NonNull Call<DefaultPayload> call, @NonNull Throwable t) {
        Snackbar.make(binding.getRoot(), "Something went wrong. Try again later.", Snackbar.LENGTH_LONG).show();
    }
}

File: app/src/main/java/com/example/tetrisapp/ui/fragment/CreateLobbyFragment.kt
package com.example.tetrisapp.ui.fragment

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation.findNavController
import com.example.tetrisapp.R
import com.example.tetrisapp.databinding.FragmentCreateLobbyBinding
import com.example.tetrisapp.model.remote.response.DefaultPayload
import com.example.tetrisapp.ui.activity.MainActivity
import com.example.tetrisapp.util.FirebaseTokenUtil
import com.example.tetrisapp.util.OnTouchListener
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.gson.*
import kotlinx.coroutines.*
import kotlin.math.roundToInt

class CreateLobbyFragment : Fragment() {
    private lateinit var binding: FragmentCreateLobbyBinding
    private var token: String? = null
    private var call: HttpResponse? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCreateLobbyBinding.inflate(inflater, container, false)
        FirebaseTokenUtil.getFirebaseToken { token: String? ->
            if (token == null) findNavController(binding.root).navigate(R.id.action_createLobbyFragment_to_signUpFragment)
            this.token = token
        }
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initOnClickListeners()
    }

    override fun onDestroy() {
        super.onDestroy()
        call?.cancel()
    }

    private fun initOnClickListeners() {
        binding.chipHighScore.setOnCheckedChangeListener { _, checked ->
            binding.linearLayout3.visibility = if (checked) View.VISIBLE else View.GONE
        }

        binding.btnBack.setOnClickListener(OnTouchListener(requireActivity() as MainActivity) {
            findNavController(binding.root).popBackStack()
        })
        binding.btnResetSettings.setOnClickListener(OnTouchListener(requireActivity() as MainActivity) {
            binding.countdownSlider.value = 3.0f
            binding.playerLimitSlider.value = 2.0f
            binding.switchEnablePause.isChecked = false
        })
        binding.btnCreateLobby.setOnClickListener(OnTouchListener(requireActivity() as MainActivity) {
            val countdown = binding.countdownSlider.value.roundToInt()
            val playerLimit = binding.playerLimitSlider.value.roundToInt()
            val timer = binding.timerSlider.value.roundToInt()
            val enablePause = binding.switchEnablePause.isChecked
            val gameMode = if (binding.chipHighScore.isChecked) "highestScore" else "battleRoyale"
            startLoading()

            sendRequest(countdown, playerLimit, timer, enablePause, gameMode)
        })
    }

    private fun sendRequest(
        countdown: Int,
        playerLimit: Int,
        timer: Int,
        enablePause: Boolean,
        gameMode: String
    ) {
        val coroutineExceptionHandler = CoroutineExceptionHandler{_, t ->
            Log.e(TAG, t.localizedMessage ?: "")
            finishLoading()
        }

        lifecycleScope.launch(Dispatchers.IO + coroutineExceptionHandler) {
            HttpClient() {
                install(ContentNegotiation) {
                    gson()
                }
            }.use { client ->
                call = client.post(getString(R.string.api_url) + "lobby/create") {
                    setBody(
                        mapOf(
                            "idToken" to token,
                            "countdown" to countdown,
                            "limit" to playerLimit,
                            "enablePause" to enablePause,
                            "timer" to timer,
                            "gameMode" to gameMode
                        )
                    )
                    contentType(ContentType.Application.Json)
                }
                call?.let { response ->
                    val body: DefaultPayload<String> = response.body()
                    finishLoading()
                    if (response.status.value == 401) {
                        findNavController(binding.root).navigate(R.id.action_createLobbyFragment_to_signUpFragment)
                    }
                    if (response.status.isSuccess() && body.success) {
                        withContext(Dispatchers.Main) {
                            val action = CreateLobbyFragmentDirections.actionCreateLobbyFragmentToLobbyFragment()
                            action.inviteCode = body.data ?: ""
                            findNavController(binding.root).navigate(action)
                        }
                    } else {
                        throw(Exception(body.message))
                    }
                }
            }
        }
    }


    private fun startLoading() {
        lifecycleScope.launch(Dispatchers.Main) {
            binding.countdownSlider.isEnabled = false
            binding.playerLimitSlider.isEnabled = false
            binding.switchEnablePause.isEnabled = false
            binding.btnResetSettings.isEnabled = false
            binding.btnCreateLobby.isEnabled = false
            binding.progressBar.visibility = View.VISIBLE
        }
    }

    private fun finishLoading() {
        lifecycleScope.launch(Dispatchers.Main) {
            binding.countdownSlider.isEnabled = true
            binding.playerLimitSlider.isEnabled = true
            binding.switchEnablePause.isEnabled = true
            binding.btnResetSettings.isEnabled = true
            binding.btnCreateLobby.isEnabled = true
            binding.progressBar.visibility = View.GONE
        }
    }

    companion object {
        private const val TAG = "CreateLobbyFragment"
    }
}